### PR TITLE
chore: update QTT framework & tests to support improvements on left/outer joins

### PR DESCRIPTION
**Pull request by spena**
### Description 
_What behavior do you want to change, why, how does your patch achieve the changes?_
^ It also fixes the ksql/master branch

### Testing done 
_Describe the testing strategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java
Comment (mjsax) [inline ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java:171]: > which uses the same value format on both left/right joins.

As discussed in person. I this actually correct, ie, can we rely on it? Do we have test for mix value formats for stream-stream joins?

File: ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TopicInfoCache.java
Comment (mjsax) [inline ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TopicInfoCache.java:111]: For my own education: what is this `cache` thing? And won't we need to get `TopicInfo` from the cache again, after we refreshed it? What does `refresh()` do internally (ie, how does a refresh work)?

File: ksqldb-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java
Comment (spena) [reply ksqldb-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java]: It was more than 100 columns. I just imported the static InternalConfig to have all in one line. The null check is already done in the constructor.

File: ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java
Comment (spena) [reply ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java]: I changed this code. If the multiple values do not have the same value format, then it will throw an exception. I left a comment about the zero cases too. I cannot throw with zero cases.

File: ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java
Comment (spena) [reply ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java]: I'm not sure if add a test for this because it should be supported. But the QTT is the only client using the `getTopicInfo` and had the issue with the multiple value formats. I added code just to verify multiple values have the same format. I don't think the `getTopicInfo` will be called by anyone else. 

We should refactor the code, though, so it is only used by testing.

File: ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TopicInfoCache.java
Comment (spena) [reply ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TopicInfoCache.java:111]: The `cache.getUnchecked(topicName)` will call the `load(final String topicName)` internally if no items exist in the cache. After that is called, consecutive calls to `getUnchecked` will use the item in the cache, and `load` will not be called any more until the item has expired. 

The problem I am fixing is that sometimes, the first record found in the shared outer changelog topic has a null value. This causes the `getTopicInfo` to return a `NO_VALUE_SCHEMA_FOUND`. When that info is obtained from the `load` function, then it is kept in the cache for some time. I know that the null value is just something that may not happen too often, so a new record in the chanelog topic should allow the TrackSerde to detect the value format. However, the cache will not have the new format until the item expires, so in cases of the NONE value found, I refresh the cache so it triggers the `load` method again in case the value format was detected.

File: ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java
Comment (mjsax) [inline ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java:170]: > is only used for QTT which always executes joins with same value formats

This seems to be the way the test are effectively written but not a fundamental limitation. At it seems QTT should be able to test all valid queries and joining two streams with different value formats seems to be a valid query that we support atm, and should keep supporting?

**Review by mjsax — APPROVED** [review]
From my side LGTM. Might be good to get a second review from @agavra or @vcrfxia though, to verify.

File: ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java
Comment (agavra) [reply ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java:170]: +1 seems like something we should support and test. if there's a bug here then there might be real bugs in the code somewhere. out of scope for this PR but can you create a ticket tracking it?

File: ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TopicInfoCache.java
Comment (agavra) [reply ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TopicInfoCache.java:111]: Is that the only time that it can be `None`? Otherwise we risk always refreshing the cache and getting the same result (I think so, but worth double checking)

**Review by agavra — APPROVED** [review]
LGTM! thanks

File: ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java
Comment (vcrfxia) [reply ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java:170]: +1 to the confusion above. Is it true that this code only affects the intermediate topic value formats written into the historic plan spec files, and that the lack of support for differing value formats in stream-stream joins introduced here is only limited to what gets written into the spec, rather than our ability to run such tests and verify inputs and outputs (assuming we were to remove the error thrown on line 177 below)?

File: ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java
Comment (vcrfxia) [inline ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java:177]: Assuming my understanding above is correct, can we update this error message to be more clear about what the problem is? If this error is hit, that means that someone wrote a stream-stream join test with differing value formats, and that this is not currently supported by the test framework. It'd also be good to link to the issue that @agavra proposed we file to track a fix for this in the error message, as a way to provide additional context. (Without this clarification, I'd be really confused if I wrote a test I expected to pass and then hit this.)

**Review by vcrfxia — COMMENTED** [review]
Thanks @spena . Can we update the PR description to summarize what this PR is about? It'd be great to link to the Streams change that the changes in this PR are addressing as I don't think this PR is understandable by anyone without the relevant context.

File: ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java
Comment (mjsax) [reply ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java:177]: I am still wondering how complex it would be to lift this restriction in this PR. If too complex, I am also happy if we address it in a follow up PR, but given the new stream-stream-join requirement, it might be worth to actually add a stream-stream join QTT test with different value format now...

**Comment by vcrfxia** [comment]
We've decided to disable the stream-stream left/outer join enhancements in ksqlDB until the grace period is configurable, so this PR is not needed to unblock master anymore. (https://github.com/confluentinc/ksql/pull/7569 will be merged instead.) The changes in this PR are still useful for when we eventually do want to enable the enhancements since we'll need the updates to the QTT framework in this PR at that time. @spena do you prefer to leave this PR open and update it at that time, or close it in favor of opening a new one later? I'll leave that up to you.

**Comment by spena** [comment]
This PR was implemented in https://github.com/confluentinc/ksql/pull/7695

**Pull request closed without merging**
